src/DDTrace/Integrations/Curl/CurlIntegration.php

```diff
@@ -50,5 +57,52 @@ public static function load()
             $scope->close();
             return $result;
         });
+
+        dd_trace('curl_setopt', function ($ch, $option, $value) use ($globalConfig) {
+            // Note that curl_setopt with option CURLOPT_HTTPHEADER overwrite data instead of appending it if called
+            // multiple times on the same resource.
+            if ($option === CURLOPT_HTTPHEADER
+                    && $globalConfig->isDistributedTracingEnabled()
+                    && is_array($value)
+            ) {
+                // Storing data to be used during exec as it cannot be retrieved at then.
+                ArrayKVStore::putForResource($ch, 'http_headers', $value);
+            }
+
+            return curl_setopt($ch, $option, $value);
+        });
+
+        dd_trace('curl_setopt_array', function ($ch, $options) use ($globalConfig) {
+            // Note that curl_setopt with option CURLOPT_HTTPHEADER overwrite data instead of appending it if called
+            // multiple times on the same resource.
+            if ($globalConfig->isDistributedTracingEnabled()
+                    && array_key_exists(CURLOPT_HTTPHEADER, $options)
+            ) {
+                // Storing data to be used during exec as it cannot be retrieved at then.
+                ArrayKVStore::putForResource($ch, 'http_headers', $options[CURLOPT_HTTPHEADER]);
+            }
+
+            return curl_setopt_array($ch, $options);
+        });
+    }
+
+    /**
+     * @param resource $ch
+     */
+    public static function injectDistributedTracingHeaders($ch)
+    {
+        if (!Configuration::instance()->isDistributedTracingEnabled()) {
+            return;
+        }
+
+        $currentHttpHeaders = ArrayKVStore::getForResource($ch, 'http_headers', []);
+        if (is_array($currentHttpHeaders)) {
+            $tracer = GlobalTracer::get();
+            $context = $tracer->getActiveSpan()->getContext();
+            $carrier = Headers::colonSeparatedValuesToHeadersMap($currentHttpHeaders);
+            $tracer->inject($context, HTTP_HEADERS, $carrier);
```

Comment:
I think I may not follow here.
The problem I had is that different libraries handle headers differently. So for example, in the case of duplicate headers you may receive `['headers1: value1, value2']` for curl and `['header1' => ['value1', 'value2']]` in Guzzle. So would an additional propagator resolve this problem? :(

Reply:
Then possible two additional propagators might be in order. 
My point is however that in case of curl we can expect values like:
`[['header' => 'value1"], ['header' => 'value2']]` so converting the parameters first to associative array
means that we will lose `['header' => 'value']` when converting back array of tuples

Reply:
> My point is however that in case of curl we can expect values like
Ah ok I did not know this. Looking at them it looked like I only could get them in the form `['header: value1, value2']`, but let me check.